### PR TITLE
Continuous deployment

### PR DESCRIPTION
**Pull request by fpoli**
In this PR:
* Fix the CI script, to actually test Prusti on Windows and MacOS and not 3 times on Ubuntu. :disappointed:
* Add a GitHub workflow to build and upload Prusti artifacts to a new GitHub release each time a commit is pushed to the master branch.

### CODE REVIEW

**Comment by fpoli** [comment]
Please give a look at the `deploy.yml`, as I can't test it locally.

**Comment by fpoli** [comment]
Could someone with MacOS debug why Prusti doesn't compile on that OS?

**Comment by Aurel300** [comment]
It's strange, it compiled for me on Mac fine so far. I guess the difference is due to dependencies or the new toolchain?

**Comment by fpoli** [comment]
> It's strange, it compiled for me on Mac fine so far. I guess the difference is due to dependencies or the new toolchain?

Yes, it might be due to this change in a dependency: https://github.com/jni-rs/jni-rs/pull/230

File: x.py [hunk 1]
```diff
@@ -77,7 +77,7 @@ def get_mac_env():
     if os.path.exists(java_home):
         ld_library_path = None
         for root, _, files in os.walk(java_home):
-            if 'libjvm.dylib' in files:
+            if 'libjli.dylib' in files:
```
Comment (Aurel300) [inline x.py:80]: Also the error message on line 84.

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+          release_name: Release ${{ github.ref }}
+          draft: false
+          prerelease: false
+      - name: Upload release asset for Ubuntu
+        uses: actions/upload-release-asset@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ steps.create_release.outputs.upload_url }}
+          asset_path: ./prusti-release-ubuntu./prustizip
```
Comment (Aurel300) [inline .github/workflows/deploy.yml]:
```suggestion
          asset_path: ./prusti-release-ubuntu/prusti.zip
```

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+          asset_name: prusti-release-ubuntu.zip
+          asset_content_type: application/zip
+      - name: Upload release asset for Windows
+        uses: actions/upload-release-asset@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ steps.create_release.outputs.upload_url }}
+          asset_path: ./prusti-release-windows/prusti.zip
```
Comment (Aurel300) [inline .github/workflows/deploy.yml]:
```suggestion
          asset_path: ./prusti-release-windows/prusti.zip
```

File: .github/workflows/deploy.yml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+          asset_name: prusti-release-windows.zip
+          asset_content_type: application/zip
+      - name: Upload release asset for MacOS
+        uses: actions/upload-release-asset@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ steps.create_release.outputs.upload_url }}
+          asset_path: ./prusti-release-macos.z/prustiip
```
Comment (Aurel300) [inline .github/workflows/deploy.yml]:
```suggestion
          asset_path: ./prusti-release-macos/prusti.zip
```

(What happened here :D)

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    steps:
+      - name: Download all Prusti artifacts
+        uses: actions/download-artifact@v2
+      - name: Package Prusti artifacts
+        shell: bash
+        run: |
+          for os in ubuntu windows macos
+          do
+            echo "Package Prusti artifact for $os"
+            cd prusti-release-$os-latest
```
Comment (Aurel300) [inline .github/workflows/deploy.yml]:
```suggestion
            cd prusti-release-$os
```

IIUC this has to be the same as the ones from upload artifact?

File: .github/workflows/deploy.yml (hunk 3 above)
Comment (fpoli) [reply .github/workflows/deploy.yml]: What's the difference?

File: .github/workflows/deploy.yml (hunk 5 above)
Comment (fpoli) [reply .github/workflows/deploy.yml]: Yes, it has to be the same. Line 61 doesn't have the `-latest` suffix.

File: .github/workflows/deploy.yml (hunk 3 above)
Comment (Aurel300) [reply .github/workflows/deploy.yml]: Sorry, I added `-latest` to all of the `asset_path`s before I realised that's probably wrong, then I edited my suggestions and didn't realise that this one was now correct.

File: .github/workflows/deploy.yml (hunk 5 above)
Comment (Aurel300) [reply .github/workflows/deploy.yml]: Right, but doesn't this have to match line 38? (Which also doesn't have the suffix?)

Comment (fpoli) [reply .github/workflows/deploy.yml]: > Right, but doesn't this have to match line 38? 

Yes.

> (Which also doesn't have the suffix?)

`matrix.os` iterates over line 16, so the suffix should be there.

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            cd ..
+          done
+      - name: Create GitHub release
+        id: create_release
+        uses: actions/create-release@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          tag_name: ${{ github.ref }}
+          release_name: Release ${{ github.ref }}
```
Comment (Aurel300) [inline .github/workflows/deploy.yml:77]: So this named it `Release refs/heads/master`, which is not ideal. I guess we should change this to `github.sha` to get the commit hash?

https://github.com/viperproject/prusti-dev/releases/tag/refs%2Fheads%2Fmaster

**Pull request merged**